### ISSUE TITLE
`@clientName` shouldn't carry over 

### ISSUE THREAD

**Issue by timotheeguerin** [Azure/typespec-azure#233]
Similar to discussion we just had for `@friendlyName`, `@clietnName` shouldn't carry over `model is`

[Playground](https://azure.github.io/typespec-azure/playground?c=aW1wb3J0ICJAdHlwZXNwZWMvaHR0cCI7CtIZcmVzdNUZdmVyc2lvbmluZ8wfYXp1cmUtdG9vbHMvyCstY2xpZW50LWdlbmVyYXRvci1jb3JlIjsKCnVzaW5nIFR5cGVTcGVjLkh0dHA70BVSZXN00RVWyX3IG0HEeS5DxWpHyGkuQ29yZTsKCgpAc2VydmljZQpuYW1lc3BhY2UgQ29udG9zby5XaWRnZXRNYW5hZ2VyOwoKQOYArk5hbWUoIkN1c3RvbUZvbyIpCkByb3V0ZSgiZsQOIG9wIGZvbygpOiB2b2lkxWTHIGJhciIpIMQhYmFyIGlzxCg7&e=%40azure-tools%2Ftypespec-autorest&options=%7B%7D)

**Comment by tadelesh** [comment]
From TypeSpec doc: https://typespec.io/docs/language-basics/models#is, `is` will carry over the decorators, Does it mean it is done by compile?

**Comment by timotheeguerin** [comment]
Yeah by design decorator get carried over by the compiler and we don't have a way to opt out of that behavior but there is other way to make it behave the same.

If you check the target of the decorator node and compare to the decorator node parent and make sure they match then it should only run in the case where client name was defined there

**Comment by tadelesh** [comment]
> Yeah by design decorator get carried over by the compiler and we don't have a way to opt out of that behavior but there is other way to make it behave the same.
> 
> If you check the target of the decorator node and compare to the decorator node parent and make sure they match then it should only run in the case where client name was defined there

Will compiler provide an API to distinguish if a decorator is a direct decorator? Have `@friendlyName` in core resolve such problem? I suppose it should be done in the decorator set function.

**Comment by timotheeguerin** [comment]
yes `@friendlyName` has the same problem but we can't change it as it will be breaking, `@clientName` has the opportunity before its too late. 

And no there will not be a compiler api to do the workaround. There will be an api at some point to make decorator not carry over. The workaround shouldn't be too hard to run.

**Comment by tadelesh** [comment]
@timotheeguerin do you know how to deal with augment decorator? I could not find a way to get the exact node for the referred node.

**Comment by timotheeguerin** [comment]
Hhm yeah didn’t think of that…

**Comment by tadelesh** [comment]
When I tried to `resolveTypeReference` for a union variant, I got infinite loop. See code in this [PR](https://github.com/Azure/typespec-azure/pull/412/files).
```
@service({})
namespace N {
  union TestUnion{
    "A",
    "B": "B_v",
    string
  }
  op x(body: TestUnion): void;
}
```
```
namespace Customizations;

@@clientName(N.TestUnion.B, "BRename");
```

**Comment by timotheeguerin** [comment]
So I think this issue only matters for model and operations names so maybe could limit to that for now.
